teps/0113-produce-results-from-matrixed-pipelinetasks.md

```diff
@@ -0,0 +1,205 @@
+and [sharding][sharding]. When the fanned out `PipelineTasks` produce test outputs as `Results`, the users would 
+want to pass them to subsequent `PipelineTasks` that to process the test outputs.
+
+## Background
+
+This section describes the using `Results` in `Pipelines` - see [docs][results-docs] for further details.
+
+### String Results
+
+String `Results` is a beta feature and is referred to as `$(tasks.<pipelinetask-name>.results.<result-name>)`.
+
+String `Results` from previous `TaskRuns` can be passed individually into the `Matrix`:
+
+```yaml
+tasks:
+...
+- name: task-4
+  taskRef:
+    name: task-4
+  matrix:
+  - name: values
+    value: 
+    - (tasks.task-1.results.foo) # string
+    - (tasks.task-2.results.bar) # string
+    - (tasks.task-3.results.rad) # string
+```
+
+### Array Results
+
+Array `Results` is an alpha feature and is referred to as `$(tasks.<pipelinetask-name>.results.<result-name>[*])`.
+Array indexing is supported and is referred as `$(tasks.<pipelinetask-name>.results.<result-name>[i])` where `i`
+is the index.
+
+Array `Results` from previous `TaskRuns` can be passed into the `Matrix`:
+
+```yaml
+tasks:
+...
+- name: task-5
+  taskRef:
+    name: task-5
+  matrix:
+  - name: values
+    value: (tasks.task-4.results.foo[*]) # array
+```
+
+For further information, see [TEP-0076: Array Results][tep-0076]. 
+
+### Object Results
+
+Object `Results` is an as alpha feature and is referred to as `$(tasks.<pipelinetask-name>.results.<result-name>[*])`. 
+Object elements can be referred to as `$(tasks.<pipelinetask-name>.results.<result-name>.key)`. Object `Results` cannot
+be used in a `Matrix`.
+
+For further information, see [TEP-0075: Object Parameters and Results][tep-0075].
+
+## Proposal
+
+`Matrix` in a `PipelineTask` take `Parameters` of type `Array` to replace `Parameters` of type `String` in the `Task`. 
+We propose that `PipelineTasks` with `Matrix` produce `Results` of type `Array` that aggregate `Results` of type 
+`String` in the underlying `Tasks`.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: pr
+spec:
+  params:
+  - name: foo
+    value: 
+      - foo-1
+      - foo-2
+      - foo-3
+  pipelineSpec:
+    params:
+    - name: foo
+      type: array
+    tasks:
+    - name: pipeline-task-1 # will create 3 taskruns
+      matrix:
+      - name: foo
+        value: $(params.foo) # array parameter with 3 items   
+      taskSpec:
+        params:
+        - name: foo
+          type: string
+        results:
+        - name: bar
+          type: string          
+        steps:
+          ...
+    - name: pipeline-task-2
+      param:
+      - name: bars
+        value: $(tasks.pipeline-task-1.results.bar[*]) # array result with 3 items
+      taskSpec:
+        params:
+        - name: bars
+          type: string
+        steps:
+          ...       
```

Comment:
`pipeline-task-2` does not look right. It takes `bars` as a param of type `string` but the `pipeline-task-1` returns an `array` which is assigned to `bars`.

This is also misleading in terms of consumption. The consumer task was expecting a single value but now a list of result (images) is sent.

My understanding is, this way of producing results also does not support which image digest belongs to kaniko build of an application A vs an application B. Does it?

The kaniko task from the catalog takes the name of the image and the path to the context in the workspace. If `kaniko` task is configured to run as matrix, it cannot fan out as is with two matrix parameters `IMAGE` and `CONTEXT`. Pipeline matrix will create taskRun for every combination of `IMAGE` and `CONTEXT` rather than creating an image for an application A with context set to `app-A/` and creating an image for an application B with context set to `app-B`, etc. For example:

```
    tasks:
      - name: build-using-kaniko
        matrix:
          - name: IMAGE
            value:
              - app-A
              - app-B
              - app-C
          - name: CONTEXT
            value:
              - app-A/
              - app-B/
              - app-C
        taskRef:
          name: kaniko
```

If we take an example of `git-clone` from a catalog:

```
    tasks:
      - name: clone
        matrix:
          - name: url
            value:
              - url-A
              - url-B
              - url-C
        taskRef:
          name: git-clone
```

This matrix will produce a result with three precise URLs from where the applications was cloned. Now, which specific URL belongs to which application?

I suggest going with an alternative but instead of assigning a numerical value to key or taskRun identity, set the key as a combination of the matrix params. For example:

```
$(tasks.clone.results.url.url-A)
```

Reply:
added this option to the alternative (and other alternatives I'd considered)

Reply:
added this option to the alternative (and other alternatives I'd considered)